languages/python/oso/oso/enforcer.py

```diff
@@ -0,0 +1,122 @@
+                          raised by the ``authorize`` and ``authorize_request``
+                          methods. Should be a callable that takes four
+                          arguments: ``is_not_found``, ``actor``, ``action``,
+                          and ``resource``
+        :param read_action: The action used by the ``authorize`` method to
+                            determine whether an authorization failure should
+                            raise a ``NotFoundError`` or a ``ForbiddenError``
+        """
+        self.policy = policy
+        if get_error is None:
+            self._get_error = self._default_get_error
+        else:
+            self._get_error = get_error
+        self.read_action = read_action
+
+        super().__init__()
+
+    def _default_get_error(self, is_not_found, actor, action, resource):
+        err_class = NotFoundError if is_not_found else ForbiddenError
+        return err_class(actor, action, resource)
+
+    def authorize(self, actor, action, resource, *, check_read=True):
+        """Ensure that ``actor`` is allowed to perform ``action`` on
+        ``resource``.
+
+        If the action is permitted with an ``allow`` rule in the policy, then
+        this method returns ``None``. If the action is not permitted by the
+        policy, this method will raise an error.
+
+        The error raised by this method depends on whether the user can perform
+        the ``"read"`` action on the resource. If they cannot read the resource,
+        then a ``NotFound`` error is raised. Otherwise, a ``ForbiddenError`` is
+        raised.
+
+        :param actor: The actor performing the request.
+        :param action: The action the actor is attempting to perform.
+        :param resource: The resource being accessed.
+
+        :param check_read: If set to ``False``, a ``ForbiddenError`` is always
+        thrown on authorization failures, regardless of whether the user can
+        read the resource. Default is ``True``.
+        :type check_read: bool
+
+        """
+        if not self.policy.query_rule_once("allow", actor, action, resource):
+            is_not_found = False
+            if action == self.read_action:
+                is_not_found = True
+            elif check_read and not self.policy.query_rule_once(
+                "allow", actor, self.read_action, resource
+            ):
+                is_not_found = True
+            raise self._get_error(is_not_found, actor, action, resource)
+
+    def authorize_request(self, actor, request):
+        """Ensure that ``actor`` is allowed to send ``request`` to the server.
+
+        If there is a matching ``allow_request(user, request)`` rule in the
+        policy, then this method returns ``None``. Otherwise, this method raises
+        a ``ForbiddenError``.
+
+        :param actor: The actor performing the request.
+        :param request: The request that was sent by the user. Can be a string
```

Comment:
Let's revisit this when we do the docs. But I think we should suggest this is always the request object.